Allow Determination of Workspace Order from Output of PolarizationCorrectionWildes

cailafinn opened the pull request:
### Description of work

#### Summary of work
Adds a new `SpinStates` property to the algorithm that allows the user to select the order that the spin state (`++`, `-+`, etc.) workspaces appear in the output WorkspaceGroup.


Fixes #35067 

#### Further detail of work
- Adds the new property to the algorithm, ensuring that they appear in the right order when it has been set. 
- Adds validation to ensure that the correct number of states is provided for a given correction.
- Amends the `SpinStateValidator` to handle SpinStates (`++`) and FlipperConfigurations (`00`). 
- Updates the documentation to account for the new functionality

### To test:
1. Run the test from [this documentation](https://docs.mantidproject.org/nightly/algorithms/PolarizationCorrectionWildes-v1.html#id12)
2. Take note of the order of the workspaces in the output WorkspaceGroup. 
3. Add a line under `Flippers='00, 11'` and add `SpinStates=`--,++,-+,+-`. 
4. Run the script again. 
5. Check the output workspace group is in the order you selected. 

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

cailafinn committed:
Distinguish between spin states and flipper configuration

RE #35067

cailafinn committed:
Allow user to select ordering of output workspaces

RE #35067

cailafinn committed:
Fix output order failing for 3 inputs

RE #35067

cailafinn committed:
Correctly validate SpinStates parameter

RE #35067

cailafinn committed:
Update error messages in validator

RE #35067

cailafinn committed:
Fix missing empty spinstates check

RE #35067

cailafinn committed:
Update documentation

RE #35067

cailafinn committed:
Remove unused states

RE #35067

cailafinn committed:
Add SpinStates property to wrapper alg

RE #35067

cailafinn committed:
Mention new property in wrapper alg

RE #35067

cailafinn committed:
Refactor SpinStateValidator to generalize checking

There are a few different ways to represent the spin states of your
flipper configurations (usually in inputs) and spin states (usually in
outputs). Most notably between Wildes' method (+/-) and Fredrikze'
method (p/a). The change allows the spin state validator to be
initialised with any character to represent the on/off state of flippers
or the parallel/antiparallel state of outputs.

RE #35067

cailafinn committed:
Fix non-allowed <> tags in doxygen string

RE #35067

yusufjimoh commented on Framework/Algorithms/src/PolarizationCorrectionWildes.cpp:
```diff
@@ -528,6 +540,28 @@ API::WorkspaceGroup_sptr PolarizationCorrectionWildes::groupOutput(const Workspa
+    names.emplace_back(baseName + "_" + spinState);
+    API::AnalysisDataService::Instance().addOrReplace(names.back(), ws);
+  } else {
+    auto const &i = PolarizationCorrectionsHelpers::indexOfWorkspaceForSpinState(spinStateOrder, spinState);
```
I encountered a build issue on line 559 that seems to stem from a type mismatch in the call to indexOfWorkspaceForSpinState. The function is expecting a const std::vector<std::string>& as its first argument, but it appears that a std::string is being passed instead

yusufjimoh replied:
I encountered a build issue on line 559 that seems to stem from a type mismatch in the call to `indexOfWorkspaceForSpinState`. The function is expecting a const `std::vector<std::string>&` as its first argument, but it appears that a `std::string `is being passed instead

yusufjimoh commented:
I've noticed both `const auto` and `auto const `being used in the code. Just wanted to understand if there's a difference between the two or if there's a particular reason for using them in those specific instances.

cailafinn commented:
> I've noticed both `const auto` and `auto const `being used in the code. Just wanted to understand if there's a difference between the two or if there's a particular reason for using them in those specific instances.

No, not functionally. This is the `east const` vs `const west` debate that has been going around in C++ circles for years. It's mostly stylistic in nature. I usually use `east const` by default as it's [more consistent for what the `const` is referring to](https://hackingcpp.com/cpp/design/east_vs_west_const.html), but it's usually just best to keep to whatever convention is being used in the file you're editing. I'll have a second look over the changes and make sure they're being consistent with the file.

yusufjimoh commented:
That's really interesting to know, Thanks for the explanation.

cailafinn committed:
Fix build errors after helper changes

The way that the index finder helper works was changed in another PR.
This accounts for that by handling the optional.

The optional currently throws an error if it does not return an index,
this is not testable as the validation currently prevents providing the
spin states in such a way that this would be reached. But is implemented to
result in a helpful error rather than a segfault should code changes
result in problematic spin state strings getting through.

RE #35067

cailafinn committed:
Refactor and simplify pol index correction

RE #35067

Co-authored-by: Yusuf Jimoh  <yusufjimoh@users.noreply.github.com>

cailafinn committed:
Make const position consistent

Wildes file uses const west... *grumble*

RE #35067

yusufjimoh commented:
The build issues have been resolved. I tested the code with and without the new spin states configuration, and it functions well in both scenarios for the PolarizationCorrectionWildes and PolarizationEfficiencyCor algorithms, respectively. The output workspace group was returned in the correct order. The tests added for the new spin states functionality are thorough.

- Code is well-written
- Functions correctly
- Tests for the spin states are comprehensive
- Release note is sufficient

Everything appears to be in order.

yusufjimoh reviewed (changes requested):
I happened to notice that lines 559 and 132 might still be using auto const. not certain if that's intentional

cailafinn committed:
Make SeeAlso location consistent

RE #35067

cailafinn commented:
> I happened to notice that lines 559 and 132 might still be using auto const. not certain if that's intentional

Good catch on 559. 132 matches the rest of the file.

yusufjimoh commented:
Thanks for clarifying

sf1919 commented:
Tests have now passed so reviewing can continue on this one.

yusufjimoh commented:
All checks have been completed, and all tests are passing. Pull request approved

rbauststfc commented on Framework/Algorithms/inc/MantidAlgorithms/PolarizationCorrections/SpinStateValidator.h:
```diff
@@ -34,6 +35,13 @@ class MANTID_ALGORITHMS_DLL SpinStateValidator : public Kernel::TypedValidator<s
   static const std::string ZERO;
   static const std::string ONE;
 
+  static const std::string MINUS_PLUS;
```
It's great that we've re-defined these constants in the `PolarizationCorrectionsHelper` so that this validator is more generic. Since that change, I can't see anywhere that these spin state constants and the flipper config constants above are used now - is it possible to remove them?

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrectionWildes.cpp:
```diff
@@ -333,10 +326,14 @@ void PolarizationCorrectionWildes::init() {
+                      FlipperConfigurations::ON_OFF + ", " + FlipperConfigurations::ON_ON,
+                  flipperConfigValidator, "Flipper configurations of the input workspaces.");
+  const auto spinStateValidator =
+      std::make_shared<SpinStateValidator>(std::unordered_set<int>{0, 1, 2, 4}, true, '+', '-', true);
```
I was just wondering why the allowed number of spins was set to `{0, 1, 2, 4}` here when it looks like it's set to `{0, 2, 4}` in `PolarizationEfficiencyCor`?

cailafinn replied:
Ah, good spot. I'd misunderstood the possible outputs when I implemented the changes in Wildes, but had worked them out when I got to EfficiencyCor.

rbauststfc replied:
I realised reading through the PR that it's more complex than it first appears, so great job spotting that and figuring it all out in the end!

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrectionWildes.cpp:
```diff
@@ -528,6 +543,34 @@ API::WorkspaceGroup_sptr PolarizationCorrectionWildes::groupOutput(const Workspa
+    const auto &maybeIndex = PolarizationCorrectionsHelpers::indexOfWorkspaceForSpinState(
+        PolarizationCorrectionsHelpers::splitSpinStateString(spinStateOrder), spinState);
+    if (!maybeIndex.has_value()) {
+      throw std::invalid_argument("Required spin state (" + spinState + ") not found in spin state order (" +
```
Not necessary to change if you'd prefer not to as this validation does work to prevent invalid spin states being used and provides a clear error message. I'm just conscious that we only get to this once the full correction has been applied. For example, if I run the algorithm with spin state "-, +" then it's only at this final stage that we pick up that it's not a valid input, however I think we could potentially validate for this earlier.

cailafinn replied:
Am I right in thinking that we could fix this by only allowing the two char spin states? That's easy to set on the validator rather than doing heavier checking.

rbauststfc reviewed (commented):
Changes look really good and thanks very much for putting in the effort to make the spin state validator more generic, that will be extremely helpful for future work!

I noticed one or two minor things that I thought might be worth mentioning, although not all necessarily need changing if they were deliberate choices so feel free to go with your preference on those.

I can't seem to see any release notes on this PR, though, so it would be good to add some before merging this in.

rbauststfc reviewed (dismissed):
Sorry, I realise as this is at gatekeeper stage I'd better set to request changes to prevent merging before we have a release note.

cailafinn committed:
Remove unneeded constsants from validator

RE #35607

cailafinn committed:
Correct allowed output spins

RE #35607

cailafinn committed:
Limit to two-character spin state outputs

RE #35067

cailafinn committed:
Add a release note

RE #35067

rbauststfc committed:
Merge branch 'main' into 35067_wildes_flipper_config

rbauststfc committed:
Fix formatting of seeAlso method

rbauststfc committed:
Use new constants in PolarizationEfficienciesWildes

The flipper configuration constants have been moved from the
SpinValidator to the PolarizationCorrectionsHelpers file,
so new algorithm PolarizationEfficienciesWildes needs to be updated to use these.

rbauststfc commented:
@cailafinn, the changes on this are looking really good, I've tested the changes to the validation and that's all working well. Thanks for adding a release note too. I'm afraid [this PR](https://github.com/mantidproject/mantid/pull/37753) of mine that was merged in yesterday caused some conflicting changes and I thought it seemed a bit unfair for you to have to sort those out when you got back, so I've resolved them, I hope that's OK.

@yusufjimoh, would you be happy to do a quick re-review (i.e. look through the code to confirm you're happy with the changes since your last review, but no need to re-do manual testing at this stage)? If you're happy then you can approve the PR as normal and I will gatekeeper it.

yusuphjoluwasen reviewed (approved):
All changes look good. Thanks for adding the release note—I initially confused it with the documentation updates. I appreciate your efforts, @rbauststfc, in ensuring everything was addressed. Happy to approve the pull request.
